### DISCUSSION TITLE
Form row classes getting removed when submitting via ajax

### DISCUSSION THREAD

**Discussion by paulgoodfield** [solspace/craft-freeform#1567] [Questions & Answers] [answered]
### What happened?

Am using the default Flexbox formatting template on a form. Form is set to submit using ajax and to "Reload form with a success banner above".

When the form is submitted successfully, the form loses the `class="freeform-row"` attributes that were in place on each row div, which is messing up how the form looks.

Example:

Here's the markup for one of the rows prior to the form submitting...

```
<div class="freeform-row freeform-row">
   <div class="freeform-column-6 freeform-fieldtype-text freeform-column" data-field-container="firstName" data-field-type="text"><label class="freeform-label freeform-required" for="form-input-firstName">First Name</label>
      <input novalidate="" class="freeform-input" name="firstName" type="text" id="form-input-firstName" placeholder="Your first name" value="">
   </div>
   <div class="freeform-column-6 freeform-fieldtype-text freeform-column" data-field-container="lastName" data-field-type="text"><label class="freeform-label freeform-required" for="form-input-lastName">Last Name</label>
      <input novalidate="" class="freeform-input" name="lastName" type="text" id="form-input-lastName" placeholder="Your last name" value="">
   </div>
</div>
```

...and after the form submits, the markup for the same row in the form looks like this...

```
<div>
   <div class="freeform-column-6 freeform-fieldtype-text freeform-column" data-field-container="firstName" data-field-type="text"><label class="freeform-label freeform-required" for="form-input-firstName">First Name</label>
      <input novalidate="" class="freeform-input" name="firstName" type="text" id="form-input-firstName" placeholder="Your first name" value="">
   </div>
   <div class="freeform-column-6 freeform-fieldtype-text freeform-column" data-field-container="lastName" data-field-type="text"><label class="freeform-label freeform-required" for="form-input-lastName">Last Name</label>
      <input novalidate="" class="freeform-input" name="lastName" type="text" id="form-input-lastName" placeholder="Your last name" value="">
   </div>
</div>
```

I'm rendering the form just using the standard `{{ form.render() }}`

### How can we reproduce this?

1. Create a form that uses the 'Flexbox' formatting template
2. Make sure that forms are set to use ajax for processing
3. Make sure that forms are set to "Reload form with a success banner above" in the Success Behaviour settings
4. Render form on a page using `{{ form.render() }}`

### Freeform Edition

Pro

### Freeform Version

5.5.7

### Craft Version

4.11.3

### When did this issue start?

Fresh install of Freeform

**Comment by kjmartens** [comment]
Sorry about this issue @paulgoodfield. We've seen this issue surface sporadically and are not sure why it happens. We are going to investigate this shortly.

**Comment by alexmillman23** [comment]
I'm having a similar issue on non-ajax forms.

I am passing an ID in the `attributes` in the form render. When the submission happens and it returns back with validation errors, the attributes are stripped from the form.

**Comment by seandelaney** [comment]
@alexmillman23 @paulgoodfield 

We've identified the issue to be related to the forms `hash`, `payload` and `csrf` token being refreshed when used with caching, and/or calling `form.renderContext` before calling `form.render`. The `payload` part is causing the HTML returned upon submission to not include the `attributes` object you initially passed in.

Is caching enabled on your templates? If so, which caching type is used - e.g Blitz?

Can I please see both you're templates and any JS you use to refresh the forms hash, payload and/or CSRF tokens too?

**Comment by paulgoodfield** [comment]
@seandelaney Yes we're using Blitz on the site in question.

We're using the following JS to refresh the form hash, payload and CSRF:

```javascript
export function loadDynamicFormFields() {
	for (const form of document.querySelectorAll('[data-freeform]')) {
		const handle = form.getAttribute('data-handle');

		fetch(`/dynamic/form-fields?form=${handle}`, {
				method: "get",
			})
			.then((r) => r.json())
			.then((r) => {

				form.querySelector('input[name=formHash]').value = r.hash;

                // Update the Payload if encrypted payloads are enabled
                form.querySelector('input[name=freeform_payload]').value = r.payload;

                // Locate and update the CSRF input
                const csrf = r.csrf;
                form.querySelector('input[name=' + csrf.name + ']').value = csrf.value;	
			})
			.catch((err) => {

				console.error(err);
			});
	}
}
```
And here's the `dynamic/form-fields` twig file the JS is loading in the fetch call:

```twig
{% set form = freeform.form(craft.app.request.get('form')) %}
{{ {
    hash: form.hash,
    payload: form.payload,
    csrf: {
        name: craft.app.config.general.csrfTokenName,
        value: craft.app.request.csrfToken,
    }
}|json_encode|raw }}
```

**Comment by seandelaney** [comment, marked as answer]
@paulgoodfield 

This is exactly what I expected to see. Okay, can you please update your code to this:

```
export function loadDynamicFormFields() {
	for (const form of document.querySelectorAll('[data-freeform]')) {
		fetch('/dynamic/form-fields')
			.then(r => r.json())
			.then(r => {
                 		// Locate and update the CSRF input
				if (r.csrf) {
					const csrfInput = form.querySelector('input[name=' + r.csrf.name + ']');
					if (csrfInput && r.csrf.name && r.csrf.value) {
						csrfInput.value = r.csrf.value;
					}
                    		}
			})
			.catch(console.error);
	}
}
```

and this

```
{{ {
    csrf: {
        name: craft.app.config.general.csrfTokenName,
        value: craft.app.request.csrfToken,
    }
}|json_encode|raw }}
```

Test again but I suspect you will be good to go again!

[See latest v5 docs](https://docs.solspace.com/craft/freeform/v5/guides/templating/caching-forms/#static-page-caching--cdn--blitz-)

**Comment by kjmartens** [comment]
Freeform 5.9.15+ now includes the ability to set isolated instances of forms, which should finally resolve the issue of duplicating classes and ID's.

Check out our updated guide here:
https://docs.solspace.com/craft/freeform/v5/guides/guide/multiple-instances-of-same-form/

🙂
